zerver/lib/users.py

```diff
@@ -347,6 +348,70 @@ def access_user_by_email(
     return access_user_common(target, user_profile, allow_deactivated, allow_bots, for_admin)
 
 
+def bulk_access_user_by_email(
+    users_emails: list[str],
+    acting_user: UserProfile,
+    *,
+    allow_deactivated: bool = False,
+    allow_bots: bool = False,
+    for_admin: bool,
+) -> set[UserProfile]:
+    # using .upper() for each email and annotate(email_upper=upper("email")) in the queryset
+    # achieves the same result of doing email__iexact__in=emails but that is not supported.
+    target_emails_lower = [email.strip().upper() for email in users_emails]
+    users_profiles = (
+        UserProfile.objects.annotate(email_upper=Upper("email"))
+        .select_related(
+            "realm",
+            "realm__can_access_all_users_group",
+            "realm__can_access_all_users_group__named_user_group",
+            "realm__direct_message_initiator_group",
+            "realm__direct_message_initiator_group__named_user_group",
+            "realm__direct_message_permission_group",
+            "realm__direct_message_permission_group__named_user_group",
+            "bot_owner",
+        )
+        .filter(email_upper__in=target_emails_lower, realm=acting_user.realm)
+    )
+    valid_users_emails = {user_profile.email_upper for user_profile in users_profiles}
+    all_users_exist = all(email in valid_users_emails for email in target_emails_lower)
+
+    if not all_users_exist:
+        raise JsonableError(_("No such user with given email"))
+
+    return {
+        access_user_common(user_profile, acting_user, allow_deactivated, allow_bots, for_admin)
+        for user_profile in users_profiles
+    }
+
+
+def bulk_access_user_by_id(
+    users_ids: list[int],
+    acting_user: UserProfile,
+    *,
+    allow_deactivated: bool = False,
+    allow_bots: bool = False,
+    for_admin: bool,
+) -> set[UserProfile]:
+    users_profiles = UserProfile.objects.select_related(
+        "realm",
+        "realm__can_access_all_users_group",
+        "realm__can_access_all_users_group__named_user_group",
+        "bot_owner",
+    ).filter(id__in=users_ids, realm=acting_user.realm)
```

Comment:
So we have a bunch of naming patterns across the codebase that this does not follow.

- `user_ids` for a list of user IDs.
- `user_profiles` or `users` for a list of user objects
- `bulk_access_user_by_id` should be `bulk_access_users_by_id`

I renamed these details.

I also changed `acting_user` to be a mandatory kwarg, since that forces callers to be readable.